// @inferencesh/sdk/dist/proxy/hono.js
/**
 * @inferencesh/sdk/proxy/hono - Hono Framework Proxy Handler
 *
 * Lightweight handler for Hono, ideal for edge/serverless deployments.
 *
 * @example
 * ```typescript
 * import { Hono } from "hono";
 * import { createHandler } from "@inferencesh/sdk/proxy/hono";
 *
 * const app = new Hono();
 * app.all("/api/inference/proxy", createHandler());
 * ```
 */
import { processProxyRequest, getEnvApiKey, passthrough, headersToRecord, } from "./index";
/**
 * Creates a Hono route handler for the Inference.sh proxy.
 */
export function createHandler({ resolveApiKey = getEnvApiKey, } = {}) {
    return async (context) => {
        const responseHeaders = new Headers();
        const url = new URL(context.req.url);
        return processProxyRequest({
            framework: "hono",
            method: context.req.method,
            body: async () => {
                try {
                    return await context.req.text();
                }
                catch {
                    return undefined;
                }
            },
            headers: () => headersToRecord(context.req.raw.headers),
            header: (name) => context.req.header(name),
            query: (name) => url.searchParams.get(name) ?? undefined,
            setHeader: (name, value) => responseHeaders.set(name, value),
            error: (status, data) => new Response(JSON.stringify(data), {
                status,
                headers: {
                    "Content-Type": "application/json",
                    ...Object.fromEntries(responseHeaders.entries()),
                },
            }),
            respond: passthrough,
            apiKey: resolveApiKey,
        });
    };
}
/** @deprecated Use createHandler */
export const createRouteHandler = createHandler;
