// @inferencesh/sdk/dist/api/apps.js
/**
 * Apps API
 */
export class AppsAPI {
    constructor(http) {
        this.http = http;
    }
    /**
     * List apps with cursor-based pagination
     */
    async list(params) {
        return this.http.request('post', '/apps/list', { data: params });
    }
    /**
     * Get an app by ID
     */
    async get(appId) {
        return this.http.request('get', `/apps/${appId}`);
    }
    /**
     * Get an app by version ID
     */
    async getByVersionId(appId, versionId) {
        return this.http.request('get', `/apps/${appId}/versions/${versionId}`);
    }
    /**
     * Create a new app
     */
    async create(data) {
        return this.http.request('post', '/apps', { data });
    }
    /**
     * Update an app
     */
    async update(appId, data) {
        return this.http.request('post', `/apps/${appId}`, { data });
    }
    /**
     * Delete an app
     */
    async delete(appId) {
        return this.http.request('delete', `/apps/${appId}`);
    }
    /**
     * Duplicate an app
     */
    async duplicate(appId) {
        return this.http.request('post', `/apps/${appId}/duplicate`);
    }
    /**
     * List app versions
     */
    async listVersions(appId, params) {
        return this.http.request('post', `/apps/${appId}/versions/list`, { data: params });
    }
    /**
     * Transfer app ownership to another team
     */
    async transferOwnership(appId, newTeamId) {
        return this.http.request('post', `/apps/${appId}/transfer`, { data: { team_id: newTeamId } });
    }
    /**
     * Update app visibility
     */
    async updateVisibility(appId, visibility) {
        return this.http.request('post', `/apps/${appId}/visibility`, { data: { visibility } });
    }
    /**
     * Get an app by namespace and name (e.g., "inference/claude-haiku")
     */
    async getByName(name) {
        return this.http.request('get', `/apps/${name}`);
    }
    /**
     * Get app license record
     */
    async getLicense(appId) {
        return this.http.request('get', `/apps/${appId}/license`);
    }
    /**
     * Save app license
     */
    async saveLicense(appId, license) {
        return this.http.request('post', `/apps/${appId}/license`, { data: { license } });
    }
    /**
     * Set the current (active) version of an app
     */
    async setCurrentVersion(appId, versionId) {
        return this.http.request('post', `/apps/${appId}/current-version`, { data: { version_id: versionId } });
    }
}
export function createAppsAPI(http) {
    return new AppsAPI(http);
}
