// @inferencesh/sdk/dist/api/engines.js
/**
 * Engines API
 */
export class EnginesAPI {
    constructor(http) {
        this.http = http;
    }
    /**
     * List engines with cursor-based pagination
     */
    async list(params) {
        return this.http.request('post', '/engines/list', { data: params });
    }
    /**
     * Get an engine by ID
     */
    async get(engineId) {
        return this.http.request('get', `/engines/${engineId}`);
    }
    /**
     * Get engines for specific resources (apps/agents)
     */
    async getForResources(request) {
        return this.http.request('post', '/engines/resources', { data: request });
    }
    /**
     * Create a new engine
     */
    async create(data) {
        return this.http.request('post', '/engines', { data });
    }
    /**
     * Update an engine
     */
    async update(engineId, data) {
        return this.http.request('post', `/engines/${engineId}`, { data });
    }
    /**
     * Delete an engine
     */
    async delete(engineId) {
        return this.http.request('delete', `/engines/${engineId}`);
    }
    /**
     * Stream engine updates
     */
    stream(engineId) {
        return this.http.createEventSource(`/engines/${engineId}/stream`);
    }
    /**
     * Stop an engine
     */
    async stop(engineId) {
        return this.http.request('post', `/engines/${engineId}/stop`);
    }
    /**
     * Restart an engine
     */
    async restart(engineId) {
        return this.http.request('post', `/engines/${engineId}/restart`);
    }
    /**
     * Update engine visibility
     */
    async updateVisibility(engineId, visibility) {
        return this.http.request('post', `/engines/${engineId}/visibility`, { data: { visibility } });
    }
    /**
     * Transfer engine ownership
     */
    async transferOwnership(engineId, newTeamId) {
        return this.http.request('post', `/engines/${engineId}/transfer`, { data: { team_id: newTeamId } });
    }
}
export function createEnginesAPI(http) {
    return new EnginesAPI(http);
}
