// @inferencesh/sdk/dist/api/flow-runs.js
/**
 * Flow Runs API
 */
export class FlowRunsAPI {
    constructor(http) {
        this.http = http;
    }
    /**
     * List flow runs with cursor-based pagination
     */
    async list(params) {
        return this.http.request('post', '/flowruns/list', { data: params });
    }
    /**
     * Get a flow run by ID
     */
    async get(flowRunId) {
        return this.http.request('get', `/flowruns/${flowRunId}`);
    }
    /**
     * Create a new flow run
     */
    async create(flowId, input) {
        return this.http.request('post', '/flowruns', { data: { flow: flowId, input } });
    }
    /**
     * Clone a flow run
     */
    async clone(flowRunId) {
        return this.http.request('post', `/flowruns/${flowRunId}/clone`);
    }
    /**
     * Update a flow run
     */
    async update(flowRunId, data) {
        return this.http.request('post', `/flowruns/${flowRunId}`, { data });
    }
    /**
     * Cancel a flow run
     */
    async cancel(flowRunId) {
        return this.http.request('post', `/flowruns/${flowRunId}/cancel`);
    }
    /**
     * Stream flow run updates
     */
    stream(flowRunId) {
        return this.http.createEventSource(`/flowruns/${flowRunId}/stream`);
    }
    /**
     * Stream task updates for a flow run
     */
    streamTasks(flowRunId) {
        return this.http.createEventSource(`/flowruns/${flowRunId}/tasks/stream`);
    }
    /**
     * Update flow run visibility
     */
    async updateVisibility(flowRunId, visibility) {
        return this.http.request('post', `/flowruns/${flowRunId}/visibility`, { data: { visibility } });
    }
}
export function createFlowRunsAPI(http) {
    return new FlowRunsAPI(http);
}
