// @inferencesh/sdk/dist/api/flows.js
/**
 * Flows API
 */
export class FlowsAPI {
    constructor(http) {
        this.http = http;
    }
    /**
     * List flows with cursor-based pagination
     */
    async list(params) {
        return this.http.request('post', '/flows/list', { data: params });
    }
    /**
     * Get a flow by ID
     */
    async get(flowId) {
        return this.http.request('get', `/flows/${flowId}`);
    }
    /**
     * Create a new flow
     */
    async create(name) {
        return this.http.request('post', '/flows', { data: { name } });
    }
    /**
     * Update a flow
     */
    async update(flowId, data) {
        return this.http.request('post', `/flows/${flowId}`, { data });
    }
    /**
     * Delete a flow
     */
    async delete(flowId) {
        return this.http.request('delete', `/flows/${flowId}`);
    }
    /**
     * Duplicate a flow
     */
    async duplicate(flowId) {
        return this.http.request('post', `/flows/${flowId}/duplicate`);
    }
    /**
     * List flow versions
     */
    async listVersions(flowId, params) {
        return this.http.request('post', `/flows/${flowId}/versions/list`, { data: params });
    }
    /**
     * Create an app from a flow
     */
    async createApp(flowId) {
        return this.http.request('post', `/flows/${flowId}/app`);
    }
    /**
     * Transfer flow ownership to another team
     */
    async transferOwnership(flowId, newTeamId) {
        return this.http.request('post', `/flows/${flowId}/transfer`, { data: { team_id: newTeamId } });
    }
    /**
     * Stream flow updates
     */
    stream(flowId) {
        return this.http.createEventSource(`/flows/${flowId}/stream`);
    }
}
export function createFlowsAPI(http) {
    return new FlowsAPI(http);
}
