// @inferencesh/sdk/dist/api/sessions.js
/**
 * Sessions API - namespaced session operations.
 */
/**
 * Sessions API for managing stateful worker leases.
 *
 * @example
 * ```typescript
 * const client = new Inference({ apiKey: '...' });
 *
 * // Get session info
 * const info = await client.sessions.get('sess_abc123');
 * console.log(info.status);
 *
 * // List sessions
 * const sessions = await client.sessions.list();
 *
 * // Keep session alive
 * await client.sessions.keepalive('sess_abc123');
 *
 * // End session
 * await client.sessions.end('sess_abc123');
 * ```
 */
export class SessionsAPI {
    constructor(http) {
        this.http = http;
    }
    /**
     * Get information about a session.
     *
     * @param sessionId - The session ID
     * @returns Session information
     * @throws SessionNotFoundError if session doesn't exist
     */
    async get(sessionId) {
        return this.http.request('get', `/sessions/${sessionId}`);
    }
    /**
     * List all sessions for the current user/team.
     *
     * @returns List of session information
     */
    async list() {
        const data = await this.http.request('get', '/sessions');
        return data || [];
    }
    /**
     * Extend session expiration time.
     *
     * Each call resets the expiration timer (sliding window).
     *
     * @param sessionId - The session ID
     * @returns Updated session information
     * @throws SessionNotFoundError if session doesn't exist
     * @throws SessionExpiredError if session has expired
     * @throws SessionEndedError if session was ended
     */
    async keepalive(sessionId) {
        return this.http.request('post', `/sessions/${sessionId}/keepalive`);
    }
    /**
     * End a session and release the worker.
     *
     * @param sessionId - The session ID
     * @throws SessionNotFoundError if session doesn't exist
     */
    async end(sessionId) {
        await this.http.request('delete', `/sessions/${sessionId}`);
    }
}
/**
 * Create a new SessionsAPI instance.
 */
export function createSessionsAPI(http) {
    return new SessionsAPI(http);
}
