// @inferencesh/sdk/dist/api/tasks.js
import { StreamableManager } from '../http/streamable';
import { PollManager } from '../http/poll';
import { TaskStatusCompleted, TaskStatusFailed, TaskStatusCancelled, } from '../types';
import { parseStatus } from '../utils';
//TODO: This is ugly...
function stripTask(task) {
    return {
        ...task,
        id: task.id,
        created_at: task.created_at,
        updated_at: task.updated_at,
        input: task.input,
        output: task.output,
        logs: task.logs,
        status: task.status,
        session_id: task.session_id,
    };
}
/**
 * Tasks API
 */
export class TasksAPI {
    constructor(http) {
        this.http = http;
    }
    /**
     * List tasks with cursor-based pagination
     */
    async list(params) {
        return this.http.request('post', '/tasks/list', { data: params });
    }
    /**
     * List featured tasks with cursor-based pagination
     */
    async listFeatured(params) {
        return this.http.request('get', '/tasks/featured', { params });
    }
    /**
     * Get a task by ID
     */
    async get(taskId) {
        return this.http.request('get', `/tasks/${taskId}`);
    }
    /**
     * Create and run a task
     */
    async create(data) {
        return this.http.request('post', '/apps/run', { data });
    }
    /**
     * Delete a task
     */
    async delete(taskId) {
        return this.http.request('delete', `/tasks/${taskId}`);
    }
    /**
     * Cancel a running task
     */
    async cancel(taskId) {
        return this.http.request('post', `/tasks/${taskId}/cancel`);
    }
    /**
     * Create an EventSource for streaming task updates
     */
    stream(taskId) {
        return this.http.createEventSource(`/tasks/${taskId}/stream`);
    }
    /**
     * Run a task and optionally wait for completion
     */
    async run(params, processedInput, options = {}) {
        const { onUpdate, onPartialUpdate, wait = true, autoReconnect = true, maxReconnects = 5, reconnectDelayMs = 1000, } = options;
        const task = await this.http.request('post', '/apps/run', {
            data: {
                ...params,
                input: processedInput,
            },
        });
        // Return immediately if not waiting
        if (!wait) {
            return stripTask(task);
        }
        const useStream = options.stream ?? this.http.getStreamDefault();
        if (!useStream) {
            return this.pollUntilTerminal(task, options);
        }
        // Wait for completion with optional updates via NDJSON streaming
        // Accumulate state across partial updates to preserve fields like session_id
        let accumulatedTask = { ...task };
        const { url, headers } = this.http.getStreamableConfig(`/tasks/${task.id}/stream`);
        return new Promise((resolve, reject) => {
            const streamManager = new StreamableManager({
                url,
                headers,
                onData: (data) => {
                    // Merge new data, preserving existing fields if not in update
                    accumulatedTask = { ...accumulatedTask, ...data };
                    const stripped = stripTask(accumulatedTask);
                    onUpdate?.(stripped);
                    if (parseStatus(data.status) === TaskStatusCompleted) {
                        streamManager.stop();
                        resolve(stripped);
                    }
                    else if (parseStatus(data.status) === TaskStatusFailed) {
                        streamManager.stop();
                        reject(new Error(data.error || 'task failed'));
                    }
                    else if (parseStatus(data.status) === TaskStatusCancelled) {
                        streamManager.stop();
                        reject(new Error('task cancelled'));
                    }
                },
                onPartialData: (data, fields) => {
                    // Merge partial update, preserving fields not in this update
                    accumulatedTask = { ...accumulatedTask, ...data };
                    const stripped = stripTask(accumulatedTask);
                    onPartialUpdate?.(stripped, fields);
                    if (parseStatus(data.status) === TaskStatusCompleted) {
                        streamManager.stop();
                        resolve(stripped);
                    }
                    else if (parseStatus(data.status) === TaskStatusFailed) {
                        streamManager.stop();
                        reject(new Error(data.error || 'task failed'));
                    }
                    else if (parseStatus(data.status) === TaskStatusCancelled) {
                        streamManager.stop();
                        reject(new Error('task cancelled'));
                    }
                },
                onError: (error) => {
                    reject(error);
                    streamManager.stop();
                },
            });
            streamManager.start();
        });
    }
    /** Poll GET /tasks/{id}/status until terminal, full-fetch on status change. */
    pollUntilTerminal(task, options) {
        const { onUpdate, maxReconnects = 5 } = options;
        const intervalMs = options.pollIntervalMs ?? this.http.getPollIntervalMs();
        let prevStatus = task.status;
        return new Promise((resolve, reject) => {
            const poller = new PollManager({
                pollFunction: () => this.http.request('get', `/tasks/${task.id}/status`),
                intervalMs,
                maxRetries: maxReconnects,
                onData: async (statusData) => {
                    if (statusData.status === prevStatus)
                        return;
                    prevStatus = statusData.status;
                    // Status changed — fetch full task
                    try {
                        const fullTask = await this.http.request('get', `/tasks/${task.id}`);
                        const stripped = stripTask(fullTask);
                        onUpdate?.(stripped);
                        if (parseStatus(fullTask.status) === TaskStatusCompleted) {
                            poller.stop();
                            resolve(stripped);
                        }
                        else if (parseStatus(fullTask.status) === TaskStatusFailed) {
                            poller.stop();
                            reject(new Error(fullTask.error || 'task failed'));
                        }
                        else if (parseStatus(fullTask.status) === TaskStatusCancelled) {
                            poller.stop();
                            reject(new Error('task cancelled'));
                        }
                    }
                    catch (err) {
                        poller.stop();
                        reject(err instanceof Error ? err : new Error(String(err)));
                    }
                },
                onError: (error) => {
                    reject(error);
                    poller.stop();
                },
            });
            poller.start();
        });
    }
    /**
     * Update task visibility
     */
    async updateVisibility(taskId, visibility) {
        return this.http.request('post', `/tasks/${taskId}/visibility`, { data: { visibility } });
    }
    /**
     * Feature/unfeature a task
     */
    async feature(taskId, featured) {
        return this.http.request('post', `/tasks/${taskId}/featured`, { data: { is_featured: featured } });
    }
}
export function createTasksAPI(http) {
    return new TasksAPI(http);
}
