// @inferencesh/sdk/dist/http/client.js
import { InferenceError, RequirementsNotMetException } from './errors';
import { EventSource } from 'eventsource';
/**
 * Low-level HTTP client for inference.sh API
 *
 * This client handles authentication, proxy mode, and error handling.
 * It's designed to be extended by higher-level API modules.
 */
export class HttpClient {
    constructor(config) {
        // Either apiKey, getToken, or proxyUrl must be provided
        if (!config.apiKey && !config.proxyUrl && !config.getToken) {
            throw new Error('Either apiKey, getToken, or proxyUrl is required');
        }
        this.apiKey = config.apiKey;
        this.baseUrl = config.baseUrl || 'https://api.inference.sh';
        this.proxyUrl = config.proxyUrl;
        this.getToken = config.getToken;
        this.customHeaders = { 'X-Client-Source': 'inference-sdk-js/0.5.13', ...config.headers };
        this.credentials = config.credentials || 'include';
        this.onError = config.onError;
        this.streamDefault = config.stream ?? true;
        this.pollInterval = config.pollIntervalMs ?? 2000;
    }
    /** Get the base URL */
    getBaseUrl() {
        return this.baseUrl;
    }
    /** Whether SSE streaming is the default transport (true) or polling (false) */
    getStreamDefault() {
        return this.streamDefault;
    }
    /** Get the default poll interval in ms */
    getPollIntervalMs() {
        return this.pollInterval;
    }
    /** Check if in proxy mode */
    isProxyMode() {
        return !!this.proxyUrl;
    }
    /** Resolve dynamic headers */
    resolveHeaders() {
        const resolved = {};
        for (const [key, value] of Object.entries(this.customHeaders)) {
            const val = typeof value === 'function' ? value() : value;
            if (val !== undefined) {
                resolved[key] = val;
            }
        }
        return resolved;
    }
    /** Get authorization token */
    getAuthToken() {
        if (this.getToken) {
            return this.getToken();
        }
        return this.apiKey;
    }
    /**
     * Make an HTTP request to the API
     */
    async request(method, endpoint, options = {}) {
        const doRequest = () => this.executeRequest(method, endpoint, options);
        try {
            return await doRequest();
        }
        catch (error) {
            if (this.onError) {
                return await this.onError(error, doRequest);
            }
            throw error;
        }
    }
    /**
     * Execute the actual HTTP request (internal)
     */
    async executeRequest(method, endpoint, options = {}) {
        // Build the target URL (always points to the API)
        const targetUrl = new URL(`${this.baseUrl}${endpoint}`);
        if (options.params) {
            Object.entries(options.params).forEach(([key, value]) => {
                if (value !== undefined && value !== null) {
                    // Serialize arrays and objects as JSON, primitives as strings
                    const serialized = typeof value === 'object' ? JSON.stringify(value) : String(value);
                    targetUrl.searchParams.append(key, serialized);
                }
            });
        }
        // In proxy mode, requests go to the proxy with target URL in a header
        const isProxyMode = !!this.proxyUrl;
        const fetchUrl = isProxyMode ? this.proxyUrl : targetUrl.toString();
        const headers = {
            'Content-Type': 'application/json',
            ...this.resolveHeaders(),
        };
        if (isProxyMode) {
            // Proxy mode: send target URL as header, no auth (proxy handles it)
            headers['x-inf-target-url'] = targetUrl.toString();
        }
        else {
            // Direct mode: include authorization header
            const token = this.getAuthToken();
            if (token) {
                headers['Authorization'] = `Bearer ${token}`;
            }
        }
        const fetchOptions = {
            method: method.toUpperCase(),
            headers,
            credentials: this.credentials,
        };
        if (options.data) {
            fetchOptions.body = JSON.stringify(options.data);
        }
        const response = await fetch(fetchUrl, fetchOptions);
        const responseText = await response.text();
        // Try to parse as JSON
        let data = null;
        try {
            data = JSON.parse(responseText);
        }
        catch {
            // Not JSON
        }
        // Check for HTTP errors
        if (!response.ok) {
            // Check for RequirementsNotMetException (412 with errors array)
            if (response.status === 412 && data && 'errors' in data && Array.isArray(data.errors)) {
                throw RequirementsNotMetException.fromResponse(data, response.status);
            }
            // General error handling
            let errorDetail;
            if (data && typeof data === 'object') {
                const apiData = data;
                if (apiData.error) {
                    errorDetail = typeof apiData.error === 'object' ? apiData.error.message : String(apiData.error);
                }
                else if ('message' in data) {
                    errorDetail = String(data.message);
                }
                else {
                    errorDetail = JSON.stringify(data);
                }
            }
            else if (responseText) {
                errorDetail = responseText.slice(0, 500);
            }
            throw new InferenceError(response.status, errorDetail || 'Request failed', responseText);
        }
        // Handle 204 No Content (e.g., DELETE requests)
        if (response.status === 204) {
            return undefined;
        }
        const apiResponse = data;
        if (!apiResponse?.success) {
            let errorMessage = apiResponse?.error?.message;
            if (!errorMessage) {
                errorMessage = `Request failed (success=false). Response: ${responseText.slice(0, 500)}`;
            }
            throw new InferenceError(response.status, errorMessage, responseText);
        }
        return apiResponse.data;
    }
    /**
     * Get URL and headers for NDJSON streaming.
     * Returns the full URL and auth headers needed for streamable requests.
     */
    getStreamableConfig(endpoint) {
        const targetUrl = new URL(`${this.baseUrl}${endpoint}`);
        const isProxyMode = !!this.proxyUrl;
        let url;
        const headers = { ...this.resolveHeaders() };
        if (isProxyMode) {
            const proxyUrlWithQuery = new URL(this.proxyUrl, typeof window !== 'undefined' ? window.location.origin : 'http://localhost');
            proxyUrlWithQuery.searchParams.set('__inf_target', targetUrl.toString());
            url = proxyUrlWithQuery.toString();
            headers['x-inf-target-url'] = targetUrl.toString();
        }
        else {
            url = targetUrl.toString();
            const token = this.getAuthToken();
            if (token) {
                headers['Authorization'] = `Bearer ${token}`;
            }
        }
        return { url, headers };
    }
    /**
     * Create an EventSource for SSE streaming
     * @deprecated Use getStreamableConfig() with StreamableManager instead
     */
    createEventSource(endpoint) {
        const targetUrl = new URL(`${this.baseUrl}${endpoint}`);
        const isProxyMode = !!this.proxyUrl;
        // For proxy mode: Browser EventSource can't send custom headers,
        // so append target URL as query param instead
        let fetchUrl;
        if (isProxyMode) {
            const proxyUrlWithQuery = new URL(this.proxyUrl, typeof window !== 'undefined' ? window.location.origin : 'http://localhost');
            proxyUrlWithQuery.searchParams.set('__inf_target', targetUrl.toString());
            fetchUrl = proxyUrlWithQuery.toString();
        }
        else {
            fetchUrl = targetUrl.toString();
        }
        const resolvedHeaders = this.resolveHeaders();
        return Promise.resolve(new EventSource(fetchUrl, {
            fetch: (input, init) => {
                const headers = {
                    ...init?.headers,
                    ...resolvedHeaders,
                };
                if (isProxyMode) {
                    // Proxy mode: also send target URL as header (for non-browser clients)
                    headers['x-inf-target-url'] = targetUrl.toString();
                }
                else {
                    // Direct mode: include authorization header
                    const token = this.getAuthToken();
                    if (token) {
                        headers['Authorization'] = `Bearer ${token}`;
                    }
                }
                return fetch(input, {
                    ...init,
                    headers,
                    credentials: this.credentials,
                });
            },
        }));
    }
}
/**
 * Create an HTTP client instance
 */
export function createHttpClient(config) {
    return new HttpClient(config);
}
