// @inferencesh/sdk/dist/http/poll.js
/**
 * PollManager — polling transport alternative to StreamManager.
 *
 * Used when long-lived SSE connections are unavailable (Convex actions,
 * Cloudflare Workers, restricted edge runtimes).
 */
export class PollManager {
    constructor(options) {
        this.interval = null;
        this.retryTimeout = null;
        this.consecutiveErrors = 0;
        this.isStopped = true;
        this.polling = false;
        this.options = {
            intervalMs: 2000,
            maxRetries: 5,
            retryDelayMs: 1000,
            ...options,
        };
    }
    /** Start polling. First poll is immediate, then at intervalMs. */
    start() {
        if (!this.isStopped)
            return;
        this.isStopped = false;
        this.consecutiveErrors = 0;
        this.options.onStart?.();
        this.poll(); // immediate first poll
        this.interval = setInterval(() => this.poll(), this.options.intervalMs);
    }
    /** Stop polling and clean up. */
    stop() {
        if (this.isStopped)
            return;
        this.isStopped = true;
        this.cleanup();
        this.options.onStop?.();
    }
    cleanup() {
        if (this.interval) {
            clearInterval(this.interval);
            this.interval = null;
        }
        if (this.retryTimeout) {
            clearTimeout(this.retryTimeout);
            this.retryTimeout = null;
        }
    }
    async poll() {
        if (this.isStopped || this.polling)
            return;
        this.polling = true;
        try {
            const data = await this.options.pollFunction();
            this.polling = false;
            if (this.isStopped)
                return;
            this.consecutiveErrors = 0;
            this.options.onData?.(data);
        }
        catch (err) {
            this.polling = false;
            if (this.isStopped)
                return;
            const error = err instanceof Error ? err : new Error(String(err));
            this.consecutiveErrors++;
            this.options.onError?.(error);
            if (this.consecutiveErrors >= this.options.maxRetries) {
                this.stop();
            }
        }
    }
}
