// @inferencesh/sdk/dist/http/streamable.js
/**
 * Streamable HTTP client using fetch + ReadableStream.
 * This is the NDJSON alternative to SSE/EventSource that avoids
 * the browser's ~6 connection limit per domain.
 */
/**
 * Stream NDJSON from an HTTP endpoint using fetch + ReadableStream.
 * Yields parsed JSON objects, automatically filtering heartbeats.
 */
export async function* streamable(url, options = {}) {
    const { headers = {}, body, method = body ? 'POST' : 'GET', signal, skipHeartbeats = true, } = options;
    const requestHeaders = {
        Accept: 'application/x-ndjson',
        ...headers,
    };
    if (body) {
        requestHeaders['Content-Type'] = 'application/json';
    }
    const res = await fetch(url, {
        method,
        headers: requestHeaders,
        body: body ? JSON.stringify(body) : undefined,
        signal,
    });
    if (!res.ok) {
        const text = await res.text();
        throw new Error(`HTTP ${res.status}: ${text}`);
    }
    if (!res.body) {
        throw new Error('No response body');
    }
    const reader = res.body.getReader();
    const decoder = new TextDecoder();
    let buffer = '';
    try {
        while (true) {
            const { done, value } = await reader.read();
            if (done)
                break;
            buffer += decoder.decode(value, { stream: true });
            const lines = buffer.split('\n');
            buffer = lines.pop();
            for (const line of lines) {
                if (!line.trim())
                    continue;
                const parsed = JSON.parse(line);
                // Skip heartbeats
                if (skipHeartbeats && parsed.type === 'heartbeat') {
                    continue;
                }
                // If it's a wrapped message with event/data/fields, yield the data
                // Otherwise yield the entire parsed object
                if ('data' in parsed && parsed.data !== undefined) {
                    yield parsed.data;
                }
                else {
                    yield parsed;
                }
            }
        }
        // Handle remaining buffer
        if (buffer.trim()) {
            const parsed = JSON.parse(buffer);
            if (!(skipHeartbeats && parsed.type === 'heartbeat')) {
                if ('data' in parsed && parsed.data !== undefined) {
                    yield parsed.data;
                }
                else {
                    yield parsed;
                }
            }
        }
    }
    finally {
        reader.releaseLock();
    }
}
export class StreamableManager {
    constructor(options) {
        this.abortController = null;
        this.isRunning = false;
        this.eventListeners = new Map();
        this.options = options;
    }
    /**
     * Add a listener for typed events (e.g., 'chats', 'chat_messages').
     * Used when server sends events with {"event": "eventName", "data": ...} format.
     */
    addEventListener(eventName, callback) {
        const listeners = this.eventListeners.get(eventName) || new Set();
        listeners.add(callback);
        this.eventListeners.set(eventName, listeners);
        // Return cleanup function
        return () => {
            const listeners = this.eventListeners.get(eventName);
            if (listeners) {
                listeners.delete(callback);
                if (listeners.size === 0) {
                    this.eventListeners.delete(eventName);
                }
            }
        };
    }
    async start() {
        if (this.isRunning)
            return;
        this.isRunning = true;
        this.abortController = new AbortController();
        try {
            this.options.onStart?.();
            for await (const message of streamableRaw(this.options.url, {
                headers: this.options.headers,
                body: this.options.body,
                signal: this.abortController.signal,
            })) {
                if (!this.isRunning)
                    break;
                const wrapper = message;
                // Handle typed events ({"event": "...", "data": ...})
                if (wrapper.event && this.eventListeners.has(wrapper.event)) {
                    const listeners = this.eventListeners.get(wrapper.event);
                    const eventData = wrapper.data !== undefined ? wrapper.data : message;
                    listeners.forEach(callback => callback(eventData));
                    continue;
                }
                // Check if it's a partial update
                if (typeof message === 'object' &&
                    message !== null &&
                    'data' in message &&
                    'fields' in message &&
                    Array.isArray(wrapper.fields)) {
                    if (this.options.onPartialData && wrapper.data !== undefined) {
                        this.options.onPartialData(wrapper.data, wrapper.fields);
                    }
                    else if (wrapper.data !== undefined) {
                        this.options.onData?.(wrapper.data);
                    }
                }
                else if (wrapper.data !== undefined) {
                    // Has data wrapper but no fields - unwrap
                    this.options.onData?.(wrapper.data);
                }
                else {
                    this.options.onData?.(message);
                }
            }
        }
        catch (err) {
            if (err instanceof Error && err.name === 'AbortError') {
                // Intentional stop, not an error
            }
            else {
                this.options.onError?.(err instanceof Error ? err : new Error(String(err)));
            }
        }
        finally {
            this.isRunning = false;
            this.options.onEnd?.();
        }
    }
    stop() {
        this.isRunning = false;
        this.abortController?.abort();
        this.abortController = null;
    }
}
/**
 * Raw streamable that yields the full message including wrapper.
 * Use this when you need access to event type or fields.
 */
export async function* streamableRaw(url, options = {}) {
    const { headers = {}, body, method = body ? 'POST' : 'GET', signal, skipHeartbeats = true, } = options;
    const requestHeaders = {
        Accept: 'application/x-ndjson',
        ...headers,
    };
    if (body) {
        requestHeaders['Content-Type'] = 'application/json';
    }
    const res = await fetch(url, {
        method,
        headers: requestHeaders,
        body: body ? JSON.stringify(body) : undefined,
        signal,
    });
    if (!res.ok) {
        const text = await res.text();
        throw new Error(`HTTP ${res.status}: ${text}`);
    }
    if (!res.body) {
        throw new Error('No response body');
    }
    const reader = res.body.getReader();
    const decoder = new TextDecoder();
    let buffer = '';
    try {
        while (true) {
            const { done, value } = await reader.read();
            if (done)
                break;
            buffer += decoder.decode(value, { stream: true });
            const lines = buffer.split('\n');
            buffer = lines.pop();
            for (const line of lines) {
                if (!line.trim())
                    continue;
                const parsed = JSON.parse(line);
                // Skip heartbeats
                if (skipHeartbeats && parsed.type === 'heartbeat') {
                    continue;
                }
                yield parsed;
            }
        }
        // Handle remaining buffer
        if (buffer.trim()) {
            const parsed = JSON.parse(buffer);
            if (!(skipHeartbeats && parsed.type === 'heartbeat')) {
                yield parsed;
            }
        }
    }
    finally {
        reader.releaseLock();
    }
}
