// @inferencesh/sdk/dist/http/stream.js
/**
 * Check if the parsed data is a partial data wrapper from the server.
 * The server sends partial updates in the format: { data: T, fields: string[] }
 */
function isPartialDataWrapper(parsed) {
    return (typeof parsed === 'object' &&
        parsed !== null &&
        'data' in parsed &&
        'fields' in parsed &&
        Array.isArray(parsed.fields));
}
export class StreamManager {
    constructor(options) {
        this.eventSource = null;
        this.reconnectTimeout = null;
        this.stopTimeout = null;
        this.initialConnectionAttempts = 0;
        this.isConnected = false;
        this.isStopped = false;
        this.eventListeners = new Map();
        this.options = {
            autoReconnect: true,
            maxReconnects: 5,
            reconnectDelayMs: 1000,
            ...options,
        };
    }
    /**
     * Add a listener for typed SSE events (e.g., 'chats', 'chat_messages')
     * Used when server sends events with `event: eventName` header
     */
    addEventListener(eventName, callback) {
        const listeners = this.eventListeners.get(eventName) || new Set();
        listeners.add(callback);
        this.eventListeners.set(eventName, listeners);
        // If already connected, set up the listener on the existing EventSource
        if (this.eventSource) {
            this.setupEventListener(this.eventSource, eventName);
        }
        // Return cleanup function
        return () => {
            const listeners = this.eventListeners.get(eventName);
            if (listeners) {
                listeners.delete(callback);
                if (listeners.size === 0) {
                    this.eventListeners.delete(eventName);
                }
            }
        };
    }
    setupEventListener(source, eventName) {
        source.addEventListener(eventName, (e) => {
            if (this.isStopped)
                return;
            try {
                const parsed = JSON.parse(e.data);
                const listeners = this.eventListeners.get(eventName);
                if (listeners?.size) {
                    // Extract actual data from partial wrapper if present
                    // Server may send {data: T, fields: [...]} for typed events too
                    const actualData = isPartialDataWrapper(parsed) ? parsed.data : parsed;
                    listeners.forEach(callback => callback(actualData));
                }
            }
            catch (err) {
                const error = err instanceof Error ? err : new Error('Invalid JSON');
                this.options.onError?.(error);
            }
        });
    }
    setupAllEventListeners(source) {
        this.eventListeners.forEach((_, eventName) => {
            this.setupEventListener(source, eventName);
        });
    }
    clearTimeouts() {
        if (this.reconnectTimeout) {
            clearTimeout(this.reconnectTimeout);
            this.reconnectTimeout = null;
        }
        if (this.stopTimeout) {
            clearTimeout(this.stopTimeout);
            this.stopTimeout = null;
        }
    }
    closeEventSource() {
        if (this.eventSource) {
            this.eventSource.close();
            this.eventSource = null;
        }
    }
    cleanup() {
        this.clearTimeouts();
        this.closeEventSource();
        this.isConnected = false;
        this.options.onStop?.();
    }
    stopAfter(delayMs) {
        this.clearTimeouts();
        this.stopTimeout = setTimeout(() => this.stop(), delayMs);
    }
    clearStopTimeout() {
        if (this.stopTimeout) {
            clearTimeout(this.stopTimeout);
            this.stopTimeout = null;
        }
    }
    stop() {
        this.isStopped = true;
        this.cleanup();
    }
    scheduleReconnect() {
        if (!this.options.autoReconnect || this.isStopped)
            return false;
        // If we had a successful connection before, always reconnect
        // Otherwise, check if we've exceeded initial connection attempts
        if (!this.isConnected && this.initialConnectionAttempts >= (this.options.maxReconnects ?? 5)) {
            return false;
        }
        this.reconnectTimeout = setTimeout(() => {
            if (!this.isStopped) {
                if (!this.isConnected) {
                    this.initialConnectionAttempts++;
                }
                this.connect();
            }
        }, this.options.reconnectDelayMs);
        return true;
    }
    async connect() {
        if (this.isStopped)
            return;
        this.cleanup();
        this.isStopped = false;
        try {
            const source = await this.options.createEventSource();
            if (!source || this.isStopped)
                return;
            this.eventSource = source;
            this.isConnected = true;
            this.options.onStart?.();
            // Set up typed event listeners (for events with `event: eventName` header)
            this.setupAllEventListeners(source);
            source.onmessage = (e) => {
                if (this.isStopped)
                    return;
                try {
                    const parsed = JSON.parse(e.data);
                    // Check if this is a partial data wrapper from the server
                    if (isPartialDataWrapper(parsed)) {
                        // Call onPartialData if provided, otherwise onData with extracted data
                        if (this.options.onPartialData) {
                            this.options.onPartialData(parsed.data, parsed.fields);
                        }
                        else {
                            this.options.onData?.(parsed.data);
                        }
                    }
                    else {
                        // Not a partial wrapper, treat as full data
                        this.options.onData?.(parsed);
                    }
                }
                catch (err) {
                    const error = err instanceof Error ? err : new Error("Invalid JSON");
                    this.options.onError?.(error);
                }
            };
            source.onerror = (evt) => {
                if (this.isStopped)
                    return;
                const error = new Error("Event stream error: " + evt);
                this.options.onError?.(error);
                this.cleanup();
                if (!this.scheduleReconnect()) {
                    this.stop();
                }
            };
        }
        catch (err) {
            if (this.isStopped)
                return;
            const error = err instanceof Error ? err : new Error("createEventSource failed");
            this.options.onError?.(error);
            if (!this.scheduleReconnect()) {
                this.stop();
            }
        }
    }
}
