// @inferencesh/sdk/dist/index.js
// HTTP utilities
export { HttpClient, createHttpClient } from './http/client';
export { StreamManager } from './http/stream';
export { StreamableManager, streamable, streamableRaw } from './http/streamable';
export { PollManager } from './http/poll';
export { InferenceError, RequirementsNotMetException, SessionError, SessionNotFoundError, SessionExpiredError, SessionEndedError, WorkerLostError, 
// Type guards (use these instead of instanceof for reliability)
isRequirementsNotMetException, isInferenceError, isSessionError, } from './http/errors';
// API modules
export { TasksAPI } from './api/tasks';
export { FilesAPI } from './api/files';
export { AgentsAPI, Agent } from './api/agents';
export { SessionsAPI } from './api/sessions';
export { AppsAPI } from './api/apps';
export { ChatsAPI } from './api/chats';
export { FlowsAPI } from './api/flows';
export { FlowRunsAPI } from './api/flow-runs';
export { EnginesAPI } from './api/engines';
// Tool Builder (fluent API)
export { tool, appTool, agentTool, webhookTool, internalTools, string, number, integer, boolean, enumOf, object, array, optional, } from './tool-builder';
// Status utilities (handle both int and string status values)
export { parseStatus, isTerminalStatus } from './utils';
// Types - includes TaskStatus constants and all DTOs
export * from './types';
// =============================================================================
// Main Client
// =============================================================================
import { HttpClient } from './http/client';
import { TasksAPI } from './api/tasks';
import { FilesAPI } from './api/files';
import { AgentsAPI } from './api/agents';
import { SessionsAPI } from './api/sessions';
import { AppsAPI } from './api/apps';
import { ChatsAPI } from './api/chats';
import { FlowsAPI } from './api/flows';
import { FlowRunsAPI } from './api/flow-runs';
import { EnginesAPI } from './api/engines';
/**
 * Inference.sh SDK Client
 *
 * @example
 * ```typescript
 * const client = new Inference({ apiKey: 'your-api-key' });
 * const result = await client.run({ app: 'my-app', input: { prompt: 'hello' } });
 * ```
 */
export class Inference {
    constructor(config) {
        // Handle both simple config and full HttpClientConfig
        this.http = new HttpClient({
            apiKey: 'apiKey' in config ? config.apiKey : undefined,
            baseUrl: config.baseUrl,
            proxyUrl: config.proxyUrl,
            getToken: 'getToken' in config ? config.getToken : undefined,
            headers: 'headers' in config ? config.headers : undefined,
            credentials: 'credentials' in config ? config.credentials : undefined,
            onError: 'onError' in config ? config.onError : undefined,
            stream: 'stream' in config ? config.stream : undefined,
            pollIntervalMs: 'pollIntervalMs' in config ? config.pollIntervalMs : undefined,
        });
        this.files = new FilesAPI(this.http);
        this.tasks = new TasksAPI(this.http);
        this.agents = new AgentsAPI(this.http, this.files);
        this.sessions = new SessionsAPI(this.http);
        this.apps = new AppsAPI(this.http);
        this.chats = new ChatsAPI(this.http);
        this.flows = new FlowsAPI(this.http);
        this.flowRuns = new FlowRunsAPI(this.http);
        this.engines = new EnginesAPI(this.http);
    }
    // Legacy methods for backward compatibility
    /** @internal */
    async _request(method, endpoint, options = {}) {
        return this.http.request(method, endpoint, options);
    }
    /** @internal */
    _createEventSource(endpoint) {
        return this.http.createEventSource(endpoint);
    }
    /**
     * Run a task on inference.sh
     */
    async run(params, options = {}) {
        const processedInput = await this.files.processInput(params.input);
        return this.tasks.run(params, processedInput, options);
    }
    /**
     * Upload a file
     */
    async uploadFile(data, options = {}) {
        return this.files.upload(data, options);
    }
    /**
     * Cancel a running task
     */
    async cancel(taskId) {
        return this.tasks.cancel(taskId);
    }
    /**
     * Get a task by ID
     */
    async getTask(taskId) {
        return this.tasks.get(taskId);
    }
    /**
     * Create an EventSource for streaming task updates
     */
    streamTask(taskId) {
        return this.tasks.stream(taskId);
    }
    /**
     * Create an agent for chat interactions
     */
    agent(config, options) {
        return this.agents.create(config, options);
    }
}
/**
 * Factory function for creating an Inference client (simple config)
 */
export function inference(config) {
    return new Inference(config);
}
/**
 * Create a client with extended configuration (for app integration)
 * Supports dynamic auth, custom headers, etc.
 */
export function createClient(config) {
    return new Inference(config);
}
