// @inferencesh/sdk/dist/tool-builder.js
/**
 * Tool Builder - Fluent API for defining agent tools
 */
import { ToolTypeClient, ToolTypeApp, ToolTypeAgent, ToolTypeHook } from './types';
// =============================================================================
// Schema Builders
// =============================================================================
export const string = (description) => ({ type: 'string', description });
export const number = (description) => ({ type: 'number', description });
export const integer = (description) => ({ type: 'integer', description });
export const boolean = (description) => ({ type: 'boolean', description });
export const enumOf = (values, description) => ({ type: 'string', enum: values, description });
export const object = (properties, description) => ({ type: 'object', properties, description });
export const array = (items, description) => ({ type: 'array', items, description });
export const optional = (schema) => ({ ...schema, optional: true });
// =============================================================================
// JSON Schema Generator
// =============================================================================
function toJsonSchema(params) {
    const properties = {};
    const required = [];
    for (const [key, schema] of Object.entries(params)) {
        const prop = { type: schema.type };
        if (schema.description)
            prop.description = schema.description;
        if (schema.enum)
            prop.enum = schema.enum;
        if (schema.properties) {
            const nested = toJsonSchema(schema.properties);
            prop.properties = nested.properties;
            if (nested.required.length > 0)
                prop.required = nested.required;
        }
        if (schema.items) {
            const itemSchema = toJsonSchema({ _item: schema.items });
            prop.items = itemSchema.properties['_item'];
        }
        properties[key] = prop;
        if (!schema.optional)
            required.push(key);
    }
    return { type: 'object', properties, required };
}
// =============================================================================
// Tool Builders
// =============================================================================
class ToolBuilder {
    constructor(name) {
        this.desc = '';
        this.params = {};
        this.approval = false;
        this.name = name;
    }
    describe(description) {
        this.desc = description;
        return this;
    }
    /** @deprecated Use displayName() instead */
    display(name) {
        this.displayNameValue = name;
        return this;
    }
    displayName(name) {
        this.displayNameValue = name;
        return this;
    }
    param(name, schema) {
        this.params[name] = schema;
        return this;
    }
    requireApproval() {
        this.approval = true;
        return this;
    }
}
class ClientToolBuilder extends ToolBuilder {
    /** Build schema only (use .handler() to include handler) */
    build() {
        return {
            name: this.name,
            display_name: this.displayNameValue || this.name,
            description: this.desc,
            type: ToolTypeClient,
            require_approval: this.approval || undefined,
            client: { input_schema: toJsonSchema(this.params) },
        };
    }
    /** Define handler and return ClientTool (schema + handler) */
    handler(fn) {
        return {
            schema: this.build(),
            handler: fn,
        };
    }
}
class AppToolBuilder extends ToolBuilder {
    constructor(name, appRef) {
        super(name);
        this.sessionEnabledValue = false;
        this.appRef = appRef;
    }
    /** Set one-time setup values (hidden from agent, passed on every call) */
    setup(values) {
        this.setupValues = values;
        return this;
    }
    /** Set default input values (agent can override these) */
    input(values) {
        this.inputValues = values;
        return this;
    }
    /** Set which function to call on multi-function apps */
    function(name) {
        this.functionName = name;
        return this;
    }
    /** Enable session support (agent can pass session parameter and sees session_id in output) */
    sessionEnabled() {
        this.sessionEnabledValue = true;
        return this;
    }
    build() {
        return {
            name: this.name,
            display_name: this.displayNameValue || this.name,
            description: this.desc,
            type: ToolTypeApp,
            require_approval: this.approval || undefined,
            app: {
                ref: this.appRef,
                function: this.functionName,
                session_enabled: this.sessionEnabledValue || undefined,
                setup: this.setupValues,
                input: this.inputValues,
            },
        };
    }
}
class AgentToolBuilder extends ToolBuilder {
    constructor(name, agentRef) {
        super(name);
        this.agentRef = agentRef;
    }
    build() {
        return {
            name: this.name,
            display_name: this.displayNameValue || this.name,
            description: this.desc,
            type: ToolTypeAgent,
            require_approval: this.approval || undefined,
            agent: { ref: this.agentRef },
        };
    }
}
class WebhookToolBuilder extends ToolBuilder {
    constructor(name, url) {
        super(name);
        this.url = url;
    }
    secret(key) {
        this.secretKey = key;
        return this;
    }
    build() {
        return {
            name: this.name,
            display_name: this.displayNameValue || this.name,
            description: this.desc,
            type: ToolTypeHook,
            require_approval: this.approval || undefined,
            hook: { url: this.url, secret: this.secretKey, input_schema: toJsonSchema(this.params) },
        };
    }
}
// =============================================================================
// Public API
// =============================================================================
/** Create a client tool (executed by SDK consumer) */
export const tool = (name) => new ClientToolBuilder(name);
/** Create an app tool (runs another inference app) */
export const appTool = (name, appRef) => new AppToolBuilder(name, appRef);
/** Create an agent tool (delegates to sub-agent) */
export const agentTool = (name, agentRef) => new AgentToolBuilder(name, agentRef);
/** Create a webhook tool (calls external URL) */
export const webhookTool = (name, url) => new WebhookToolBuilder(name, url);
// =============================================================================
// Internal Tools Builder
// =============================================================================
class InternalToolsBuilder {
    constructor() {
        this.config = {};
    }
    /** Enable plan tools (Create, Update, Load) */
    plan(enabled = true) {
        this.config.plan = enabled;
        return this;
    }
    /** Enable memory tools (Set, Get, GetAll) */
    memory(enabled = true) {
        this.config.memory = enabled;
        return this;
    }
    /** Enable widget tools (UI, HTML) - top-level only */
    widget(enabled = true) {
        this.config.widget = enabled;
        return this;
    }
    /** Enable finish tool - sub-agents only */
    finish(enabled = true) {
        this.config.finish = enabled;
        return this;
    }
    /** Enable all internal tools */
    all() {
        this.config.plan = true;
        this.config.memory = true;
        this.config.widget = true;
        this.config.finish = true;
        return this;
    }
    /** Disable all internal tools */
    none() {
        this.config.plan = false;
        this.config.memory = false;
        this.config.widget = false;
        this.config.finish = false;
        return this;
    }
    build() {
        return this.config;
    }
}
/** Create internal tools configuration */
export const internalTools = () => new InternalToolsBuilder();
