// @inferencesh/sdk/dist/types.js
// Code generated by tygo. DO NOT EDIT.
export const ToolTypeApp = "app"; // App tools - creates a Task
export const ToolTypeAgent = "agent"; // Sub-agent tools - creates a sub-Chat
export const ToolTypeHook = "hook"; // Webhook tools - HTTP POST to external URL
export const ToolTypeClient = "client"; // Client tools - executed by frontend
export const ToolTypeInternal = "internal"; // Internal/built-in tools (plan, memory, widget, finish)
export const ScopeGroupAgents = "agents";
export const ScopeGroupApps = "apps";
export const ScopeGroupConversations = "conversations";
export const ScopeGroupFiles = "files";
export const ScopeGroupDatastores = "datastores";
export const ScopeGroupFlows = "flows";
export const ScopeGroupProjects = "projects";
export const ScopeGroupTeams = "teams";
export const ScopeGroupBilling = "billing";
export const ScopeGroupSecrets = "secrets";
export const ScopeGroupIntegrations = "integrations";
export const ScopeGroupEngines = "engines";
export const ScopeGroupApiKeys = "apikeys";
export const ScopeGroupUser = "user";
export const ScopeGroupSettings = "settings";
export const AppCategoryImage = "image";
export const AppCategoryVideo = "video";
export const AppCategoryAudio = "audio";
export const AppCategoryText = "text";
export const AppCategoryChat = "chat";
export const AppCategory3D = "3d";
export const AppCategoryOther = "other";
export const AppCategoryFlow = "flow";
export const GPUTypeAny = "any";
export const GPUTypeNone = "none";
export const GPUTypeIntel = "intel";
export const GPUTypeNvidia = "nvidia";
export const GPUTypeAMD = "amd";
export const GPUTypeApple = "apple";
export const AppSessionStatusActive = "active";
export const AppSessionStatusEnded = "ended";
export const AppSessionStatusExpired = "expired";
export const VisibilityPrivate = "private";
export const VisibilityPublic = "public";
export const VisibilityUnlisted = "unlisted";
export const ChatStatusBusy = "busy";
export const ChatStatusIdle = "idle";
export const ChatStatusAwaitingInput = "awaiting_input";
export const ChatStatusCompleted = "completed";
export const PlanStepStatusPending = "pending";
export const PlanStepStatusInProgress = "in_progress";
export const PlanStepStatusCompleted = "completed";
export const PlanStepStatusCancelled = "cancelled";
export const ChatMessageRoleSystem = "system";
export const ChatMessageRoleUser = "user";
export const ChatMessageRoleAssistant = "assistant";
export const ChatMessageRoleTool = "tool";
export const ChatMessageStatusPending = "pending"; // Message is being generated
export const ChatMessageStatusReady = "ready"; // Message complete, tools ready for execution
export const ChatMessageStatusFailed = "failed"; // Message generation failed
export const ChatMessageStatusCancelled = "cancelled"; // Message was cancelled
export const ChatMessageContentTypeText = "text";
export const ChatMessageContentTypeReasoning = "reasoning";
export const ChatMessageContentTypeImage = "image";
export const ChatMessageContentTypeFile = "file";
export const ChatMessageContentTypeTool = "tool";
export const IntegrationTypeSlack = "slack";
export const IntegrationTypeDiscord = "discord";
export const IntegrationTypeTeams = "teams";
export const IntegrationTypeTelegram = "telegram";
export const OpEqual = "eq";
export const OpNotEqual = "neq";
export const OpIn = "in";
export const OpNotIn = "not_in";
export const OpGreater = "gt";
export const OpGreaterEqual = "gte";
export const OpLess = "lt";
export const OpLessEqual = "lte";
export const OpLike = "like";
export const OpILike = "ilike";
export const OpContains = "contains";
export const OpNotContains = "not_contains";
/**
 * Null checks
 */
export const OpIsNull = "is_null";
export const OpIsNotNull = "is_not_null";
/**
 * Empty checks (for strings)
 */
export const OpIsEmpty = "is_empty";
export const OpIsNotEmpty = "is_not_empty";
export const DeviceAuthStatusPending = "pending";
export const DeviceAuthStatusApproved = "approved";
export const DeviceAuthStatusExpired = "expired";
export const DeviceAuthStatusDenied = "denied";
export const DeviceAuthStatusValid = "valid";
export const DeviceAuthStatusInvalid = "invalid";
export const DeviceAuthStatusLoading = "loading";
export const EngineStatusRunning = "running";
export const EngineStatusPending = "pending";
export const EngineStatusDraining = "draining";
export const EngineStatusStopping = "stopping";
export const EngineStatusStopped = "stopped";
export const FlowRunStatusUnknown = 0; // 0
export const FlowRunStatusPending = 1; // 1
export const FlowRunStatusRunning = 2; // 2
export const FlowRunStatusCompleted = 3; // 3
export const FlowRunStatusFailed = 4; // 4
export const FlowRunStatusCancelled = 5; // 5
export const GraphNodeTypeUnknown = "unknown";
export const GraphNodeTypeJoin = "join";
export const GraphNodeTypeSplit = "split";
export const GraphNodeTypeExecution = "execution";
export const GraphNodeTypeResource = "resource";
export const GraphNodeTypeApproval = "approval";
export const GraphNodeTypeConditional = "conditional";
export const GraphNodeTypeFlowNode = "flow_node"; // Flow execution node
export const GraphNodeStatusPending = "pending";
export const GraphNodeStatusReady = "ready"; // Dependencies satisfied
export const GraphNodeStatusRunning = "running";
export const GraphNodeStatusCompleted = "completed";
export const GraphNodeStatusFailed = "failed";
export const GraphNodeStatusCancelled = "cancelled";
export const GraphNodeStatusSkipped = "skipped"; // Conditional branch not taken
export const GraphNodeStatusBlocked = "blocked"; // Dependency failed
export const GraphEdgeTypeDependency = "dependency"; // Blocking dependency
export const GraphEdgeTypeFlow = "flow"; // Non-blocking flow
export const GraphEdgeTypeConditional = "conditional"; // Conditional flow
export const GraphEdgeTypeExecution = "execution"; // Node → Resource execution link
export const GraphEdgeTypeParent = "parent"; // Parent resource → child resource (e.g. skill fork)
export const GraphEdgeTypeAncestor = "ancestor"; // Specific version → child resource (fork point)
export const GraphEdgeTypeDuplicate = "duplicate"; // Skill → skill with identical content hash
export const ProjectTypeAgent = "agent";
export const ProjectTypeApp = "app";
export const ProjectTypeFlow = "flow";
export const ProjectTypeOther = "other";
export const ContentSafe = "safe";
/**
 * sexual content
 */
export const ContentSexualSuggestive = "sexual_suggestive";
export const ContentSexualExplicit = "sexual_explicit";
/**
 * violence
 */
export const ContentViolenceNonGraphic = "violence_non_graphic";
export const ContentViolenceGraphic = "violence_graphic";
/**
 * gore
 */
export const ContentGore = "gore";
/**
 * other regulated content
 */
export const ContentDrugs = "drugs";
export const ContentSelfHarm = "self_harm";
export const ContentUnrated = "unrated";
export const CloudAWS = "aws";
export const CloudAzure = "azure";
export const CloudLambdaLabs = "lambdalabs";
export const CloudTensorDock = "tensordock";
export const CloudRunPod = "runpod";
export const CloudLatitude = "latitude";
export const CloudJarvisLabs = "jarvislabs";
export const CloudOblivus = "oblivus";
export const CloudPaperspace = "paperspace";
export const CloudDatacrunch = "datacrunch";
export const CloudMassedCompute = "massedcompute";
export const CloudVultr = "vultr";
export const CloudShade = "shade";
export const InstanceStatusPending = "pending";
export const InstanceStatusActive = "active";
export const InstanceStatusDeleted = "deleted";
export const TaskStatusUnknown = 0; // 0
export const TaskStatusReceived = 1; // 1
export const TaskStatusQueued = 2; // 2
export const TaskStatusDispatched = 3; // 3 - Worker assigned, task sent to worker
export const TaskStatusPreparing = 4; // 4
export const TaskStatusServing = 5; // 5
export const TaskStatusSettingUp = 6; // 6
export const TaskStatusRunning = 7; // 7
export const TaskStatusCancelling = 8; // 8 - Graceful cancellation in progress
export const TaskStatusUploading = 9; // 9
export const TaskStatusCompleted = 10; // 10
export const TaskStatusFailed = 11; // 11
export const TaskStatusCancelled = 12; // 12
export const InfraPrivate = "private";
export const InfraCloud = "cloud";
export const InfraPrivateFirst = "private_first";
export const TaskLogTypeBuild = 0; // 0
export const TaskLogTypeRun = 1;
export const TaskLogTypeServe = 2;
export const TaskLogTypeSetup = 3;
export const TaskLogTypeTask = 4;
export const TeamTypePersonal = "personal";
export const TeamTypeTeam = "team";
export const TeamTypeSystem = "system";
export const TeamRoleOwner = "owner";
export const TeamRoleAdmin = "admin";
export const TeamRoleMember = "member";
export const ToolInvocationStatusPending = "pending";
export const ToolInvocationStatusInProgress = "in_progress";
export const ToolInvocationStatusAwaitingInput = "awaiting_input"; // Waiting for user input (widgets)
export const ToolInvocationStatusAwaitingApproval = "awaiting_approval"; // Waiting for HIL approval
export const ToolInvocationStatusCompleted = "completed";
export const ToolInvocationStatusFailed = "failed";
export const ToolInvocationStatusCancelled = "cancelled";
export const UsageEventResourceTierPrivate = "private";
export const UsageEventResourceTierCloud = "cloud";
export const RoleGuest = "guest";
export const RoleUser = "user";
export const RoleAdmin = "admin";
export const RoleSystem = "system";
/**
 * Primitive node types (render literal values)
 */
export const WidgetNodeTypeText = "text";
export const WidgetNodeTypeMarkdown = "markdown";
export const WidgetNodeTypeImage = "image";
export const WidgetNodeTypeBadge = "badge";
export const WidgetNodeTypeButton = "button";
export const WidgetNodeTypeInput = "input";
export const WidgetNodeTypeSelect = "select";
export const WidgetNodeTypeCheckbox = "checkbox";
export const WidgetNodeTypeRow = "row";
export const WidgetNodeTypeCol = "col";
/**
 * Layout node types
 */
export const WidgetNodeTypeBox = "box";
export const WidgetNodeTypeSpacer = "spacer";
export const WidgetNodeTypeDivider = "divider";
export const WidgetNodeTypeForm = "form";
/**
 * Typography node types
 */
export const WidgetNodeTypeTitle = "title";
export const WidgetNodeTypeCaption = "caption";
export const WidgetNodeTypeLabel = "label";
/**
 * Control node types
 */
export const WidgetNodeTypeTextarea = "textarea";
export const WidgetNodeTypeRadioGroup = "radio-group";
export const WidgetNodeTypeDatePicker = "date-picker";
/**
 * Content node types
 */
export const WidgetNodeTypeIcon = "icon";
export const WidgetNodeTypeChart = "chart";
export const WidgetNodeTypeTransition = "transition";
/**
 * Data-bound node types (deprecated - use templates instead)
 */
export const WidgetNodeTypePlanList = "plan-list";
export const WidgetNodeTypeKeyValue = "key-value";
export const WidgetNodeTypeStatusBadge = "status-badge";
